Applications/Spire/Source/SecurityInput/SecurityInputLineEdit.cpp

```diff
@@ -68,93 +61,87 @@ const Nexus::Security& SecurityInputLineEdit::get_security() const {
-    return;
-  } else if(event->key() == Qt::Key_Enter || event->key() == Qt::Key_Return) {
-    auto current_security = ParseSecurity(text().toUpper().toStdString());
-    if(!current_security.GetSymbol().empty()) {
-      on_commit(current_security);
-    } else {
-      on_commit(m_security);
-    }
-    return;
-  } else if(event->key() == Qt::Key_Delete) {
-    on_commit({});
+  switch(event->key()) {
+    case Qt::Key_Backspace:
+      if(!text().isEmpty()) {
+        m_is_suggestion_disabled = true;
+      }
+      if(selectionLength() > 0) {
+        setText(text().remove(selectedText()));
+        m_last_key = static_cast<Qt::Key>(event->key());
+        m_securities->clear_active_item();
+        return;
+      }
+      break;
+    case Qt::Key_Enter:
+    case Qt::Key_Return:
+      {
+        auto current_security = ParseSecurity(text().toUpper().toStdString());
+        if(!current_security.GetSymbol().empty()) {
+          on_commit(current_security);
+        } else {
+          on_commit(m_security);
+        }
+        return;
+      }
+    case Qt::Key_Delete:
+      clear();
+      m_securities->hide();
+      m_is_suggestion_disabled = false;
+      return;
   }
-  QLineEdit::keyPressEvent(event);
+  TextInputWidget::keyPressEvent(event);
+  m_last_key = static_cast<Qt::Key>(event->key());
 }
 
 void SecurityInputLineEdit::paintEvent(QPaintEvent* event) {
-  QLineEdit::paintEvent(event);
+  TextInputWidget::paintEvent(event);
   if(m_is_icon_visible) {
     auto painter = QPainter(this);
     painter.drawImage(width() - ICON_WIDTH() - scale_width(8),
       height() - (height() / 2) - (ICON_HEIGHT() / 2), SEARCH_ICON());
   }
 }
 
-void SecurityInputLineEdit::resizeEvent(QResizeEvent* event) {
-  m_securities->setFixedWidth(std::max(width(), scale_width(142)));
-}
-
 void SecurityInputLineEdit::showEvent(QShowEvent* event) {
   on_text_edited();
   m_securities->setFixedWidth(std::max(width(), scale_width(142)));
 }
 
-void SecurityInputLineEdit::move_securities_list() {
-  auto x_pos = static_cast<QWidget*>(parent())->mapToGlobal(
-    geometry().bottomLeft()).x();
-  auto y_pos = static_cast<QWidget*>(parent())->mapToGlobal(
-    frameGeometry().bottomLeft()).y();
-  m_securities->move(x_pos, y_pos + 2);
-}
-
 void SecurityInputLineEdit::on_activated(const Security& security) {
-  auto item_delegate = CustomVariantItemDelegate();
-  setText(item_delegate.displayText(QVariant::fromValue(security), QLocale()));
+  setText(m_item_delegate.displayText(QVariant::fromValue(security)));
   m_security = ParseSecurity(text().toUpper().toStdString());
 }
 
 void SecurityInputLineEdit::on_commit(const Security& security) {
   m_security = security;
   Q_EMIT editingFinished();
-  m_commit_signal(security);
 }
 
 void SecurityInputLineEdit::on_text_edited() {
+  if(text().isEmpty()) {
+    m_is_suggestion_disabled = false;
+  }
   m_completions = m_model->autocomplete(text().toStdString());
   m_completions.then([=] (auto result) {
```

Comment:
Merge line 146 and 145:

    m_completions = m_model->autocomplete(...).then([=] (auto result) { ... });